[JENKINS-12875] Change default crumb name to Jenkins-Crumb

daniel-beck opened the pull request:
Supersedes #1319 rebased on `2.0`.

daniel-beck committed:
[FIXED JENKINS-12875] Change default crumb name to Jenkins-Crumb

daniel-beck commented:
FTR the original PR was 628590f.

daniel-beck commented:
> `Caused by: junit.framework.AssertionFailedError: expected:<{"a":"aaa","b":"bbb","c":{"c1":"ccc1","c2":"ccc2"},"d":{"d1":"d1","d2":"d2"},"Jenkins-Crumb":"test"}> but was:<{"a":"aaa","b":"bbb","c":{"c1":"ccc1","c2":"ccc2"},"d":{"d1":"d1","d2":"d2"}}>`

Need to investigate.

daniel-beck commented:
The test gets actual and expected wrong.

daniel-beck committed:
[JENKINS-12875] Fix test

daniel-beck commented on core/src/main/java/hudson/security/csrf/CrumbIssuer.java:
```diff
@@ -43,6 +43,7 @@
 public abstract class CrumbIssuer implements Describable<CrumbIssuer>, ExtensionPoint {
 
     private static final String CRUMB_ATTRIBUTE = CrumbIssuer.class.getName() + "_crumb";
+    public static final String DEFAULT_CRUMB_NAME = "Jenkins-Crumb";
```
`@Restricted`?

jglick replied:
Why not move it to `DefaultCrumbIssuer`?

In either case, it need not be `public`.

jglick replied:
Oh sorry, does need to be `public` for tests. But yeah could be `@Restricted(NoExternalUse.class)` I think.

daniel-beck commented:
@reviewbybees @jenkinsci/code-reviewers

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

jglick commented on core/src/main/java/hudson/security/csrf/CrumbFilter.java:
```diff
@@ -93,6 +87,21 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
 
+    private String extractCrumbFromRequest(HttpServletRequest httpRequest, String crumbFieldName) {
+        String crumb = httpRequest.getHeader(crumbFieldName);
+        if (crumb == null) {
```
BTW clearer to `return` a value as soon as you find one, then end the method with `return null`.

jglick commented on core/src/main/java/hudson/security/csrf/CrumbFilter.java:
```diff
@@ -93,6 +87,21 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
         }
     }
 
+    private String extractCrumbFromRequest(HttpServletRequest httpRequest, String crumbFieldName) {
```
Good to mark the return value `@CheckForNull`.

jglick commented on core/src/main/java/hudson/security/csrf/CrumbFilter.java:
```diff
@@ -62,17 +62,11 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
-                        break;
-                    }
-                }
+                // compatibility for clients that hard-code the default crumb name up to Jenkins 1.TODO
```
I guess this is now “…before Jenkins 2.0”? :-)

jglick commented on core/src/main/java/hudson/security/csrf/CrumbIssuerDescriptor.java:
```diff
@@ -64,7 +64,7 @@ public String getCrumbRequestField() {
     public void setCrumbRequestField(String requestField) {
         if (Util.fixEmptyAndTrim(requestField) == null) {
-            crumbRequestField = ".crumb";
+            crumbRequestField = CrumbIssuer.DEFAULT_CRUMB_NAME;
```
Weird, seems like all this stuff should really have been in `DefaultCrumbIssuer.DescriptorImpl`, not in the supertype.

daniel-beck replied:
Why would the specific implementation define the crumb name?

jglick replied:
Never mind, I got it backwards. It is actually `DefaultCrumbIssuer.DescriptorImpl.<init>` which is doing something redundant. It could just pass `null` normally to the super constructor:

```java
super(CRUMB_SALT.get(), System.getProperty("hudson.security.csrf.requestfield"));
```

so that (except for tests) the only place `DEFAULT_CRUMB_NAME` need be used is here, in `CrumbIssuerDescriptor.setCrumbRequestField`, so the constant could sensibly be moved here from `CrumbIssuer`.

jglick replied:
Even cleaner:

* inline the constant into `setCrumbRequestField`
* move the `System.getProperty` check into `setCrumbRequestField`
* create a new `protected` constructor with no crumb name, to be used from `DefaultCrumbIssuer` and `TestCrumbIssuer`
* make tests just use something like `r.jenkins.getCrumbIssuer().getCrumbRequestField()`

jglick commented on test/src/test/java/lib/form/RowSetTest.java:
```diff
@@ -27,12 +28,12 @@ public void json() throws Exception {
         public void doSubmitTest1(StaplerRequest req) throws Exception {
             JSONObject json = req.getSubmittedForm();
-            json.remove("crumb");
+            json.remove(CrumbIssuer.DEFAULT_CRUMB_NAME);
```
Huh? `crumb` != `.crumb`; was it failing to remove anything before?

jglick replied:
IOW, how could this have passed before?

jglick commented:
:bee: assuming you have tested this with nginx and confirmed that it solves the issue.

daniel-beck commented:
Works with the `ajaxExecutors` - 403 and `.crumb` header on 1.642.4, 200 and `Jenkins-Crumb` header on PR build. Many "real" forms provide the crumb in the form data, so this was the easiest one to test.

daniel-beck committed:
[JENKINS-12875] Restrict access to constant

daniel-beck commented:
Added the hopefully trivial `@Restricted` change. I implemented Jesse's https://github.com/jenkinsci/jenkins/pull/2198#discussion_r58208551 and will send a followup PR for it.
